core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OAbstractPaginatedStorage.java

```diff
@@ -1497,15 +1497,23 @@ public void preallocateRids(final OTransaction clientTx) {
 
           for (ORecordOperation txEntry : newRecords) {
             ORecord rec = txEntry.getRecord();
-
-            if (rec.isDirty()) {
-              ORecordId rid = (ORecordId) rec.getIdentity().copy();
-              ORecordId oldRID = rid.copy();
+            if (!rec.getIdentity().isPersistent()) {
+              if (rec.isDirty()) {
+                ORecordId rid = (ORecordId) rec.getIdentity().copy();
+                ORecordId oldRID = rid.copy();
+                final OCluster cluster = getClusterById(rid.getClusterId());
+                OPhysicalPosition ppos = cluster.allocatePosition(ORecordInternal.getRecordType(rec));
+                rid.setClusterId(cluster.getId());
+                rid.setClusterPosition(ppos.clusterPosition);
+                clientTx.updateIdentityAfterCommit(oldRID, rid);
+              }
+            } else {
+              ORecordId rid = (ORecordId) rec.getIdentity();
```

Comment:
Could you add a comment which explains why we need the second branch, otherwise someone can remove it because will treat it as not needed?